docs/jit/todo.org

```diff
@@ -243,8 +243,105 @@ Currently, we try to resovle the link address directly. What we should
 probably do is compute a table of linked addresses and code addresses,
 and then warn about anything that's not a link address.
 
+** Make =expr-template-compiler.pl warn for conditional-use-before-unconditional-use
+
+This is a common cause of errors, because the conditional use will
+'define' the value, and the tiler will not emit code to redefine it
+for the unconditional use. Especially dangerous is that this can
+happen in macro expansions (like write-barrier).
+
+This should probably be a hard error. How to go about checking it?
+
+First:
+- we can only get this problem in DAGs not trees
+- we can only create a DAG with named references (=$1=, =$foo=)
+- we can only have a conditional use if we have one of the following node types:
+  - WHEN (second node is conditional)
+  - IF/IFV (second and third nodes are conditional)
+  - ALL/ANY (all but first, and the 'conditionality' is relative)
+
+I'm thinking of using a stack to indicate the conditional order of
+definitions and uses.... but I'm not sure how just yet.
+
+
 * Expression Tree
 
+** Reduce tree node size to 32 bits
+
+ Tree nodes are currently 64 bits wide to allow them to coexist with
+ constant pointers. This is handy, but not really required, since we
+ could use a lookup table to get the pointers (as long as we can
+ declare pointers, for which I think we can still use the '@' sigil,
+ e.g:
+
+ #+BEGIN_EXAMPLE
+ (template: say
+    (call (const @MVM_string_say ptr_sz)
+          (arglist 2
+            (carg (tc) ptr)
+            (carg $0 ptr))
+ #+END_EXAMPLE
+
+ The @MVM_string_say pointer can be stashed in an array:
+
+ #+BEGIN_SRC C
+ static const void *MVM_jit_expr_ptrs[] = {
+    ...
+    MVM_string_say,
+    ...
+ };
+ #+END_SRC
+
+ And the pointer itself replaced by the index.
+
+ We could argue against dealing with 64 bit constants in general, but
+ unfortunately, const_i64 prevents us from doing that.... Ways of
+ dealing with that:
+
+ + A 'large constants' table per tree (into which we could copy both the
+   i64 constants and the function pointer constants)
+   + We could store this entire table in the data section, too
```

Comment:
Is https://github.com/MoarVM/MoarVM/pull/929/files#diff-ba7685d88421b4854627c06ca18b94b8R152 not doing this?

Reply:
Yes. Documentation is out of date. (The `todo.org` was an experiment to try and see if there'd be collaboration or value out of a public roadmap, but it hasn't really worked out that way).